src/Compilers/CSharp/Portable/Binder/Binder_Patterns.cs

```diff
@@ -175,11 +175,264 @@ private BoundExpression BindSwitchExpression(SwitchExpressionSyntax node, Bindin
+            {
+                info = new IndexerArgumentInfo(patternSymbol);
+                diagnostics.Add(syntax, useSiteInfo);
+                lookupResult.Free();
+                return true;
+            }
+
+            lookupResult.Free();
+            return false;
+        }
+
+        private bool PerformPatternIndexerLookup(
+            SyntaxNode syntax, TypeSymbol receiverType, LookupResult lookupResult,
+            [NotNullWhen(true)] out IndexerArgumentInfo? info, TypeSymbol indexerArgumentType, BindingDiagnosticBag diagnostics, ref CompoundUseSiteInfo<AssemblySymbol> useSiteInfo)
+        {
+            info = null;
+
+            LookupMembersInType(
+                lookupResult,
+                receiverType,
+                WellKnownMemberNames.Indexer,
+                arity: 0,
+                basesBeingResolved: null,
+                LookupOptions.Default,
+                originalBinder: this,
+                diagnose: false,
+                ref useSiteInfo);
+
+            if (lookupResult.IsMultiViable)
+            {
+                var indexerGroup = ArrayBuilder<PropertySymbol>.GetInstance();
+                foreach (Symbol symbol in lookupResult.Symbols)
+                {
+                    Debug.Assert(symbol.IsIndexer());
+                    indexerGroup.Add((PropertySymbol)symbol);
+                }
+
+                BoundExpression indexerArgumentExpr = new BoundIndexOrRangeIndexerPatternValuePlaceholder(syntax, indexerArgumentType);
+                var analyzedArguments = AnalyzedArguments.GetInstance();
+                analyzedArguments.Arguments.Add(indexerArgumentExpr);
+                BoundExpression receiver = new BoundImplicitReceiver(syntax, receiverType);
+                var overloadResolutionResult = OverloadResolutionResult<PropertySymbol>.GetInstance();
+                this.OverloadResolution.PropertyOverloadResolution(indexerGroup, receiver, analyzedArguments, overloadResolutionResult, allowRefOmittedArguments: false, ref useSiteInfo);
+                if (overloadResolutionResult.Succeeded)
+                {
+                    MemberResolutionResult<PropertySymbol> resolutionResult = overloadResolutionResult.ValidResult;
+                    PropertySymbol indexer = resolutionResult.Member;
+                    if (indexer.GetMethod is not null && IsAccessible(indexer, ref useSiteInfo))
+                    {
+                        Debug.Assert(!indexer.IsStatic);
+                        ReportDiagnosticsIfObsolete(diagnostics, indexer, syntax, hasBaseReceiver: false);
+                        CoerceArguments(resolutionResult, analyzedArguments.Arguments, diagnostics);
+                        bool isExpanded = resolutionResult.Result.Kind == MemberResolutionKind.ApplicableInExpandedForm;
+                        ImmutableArray<BoundExpression> arguments = bindDefaultArguments(indexer.GetMethod, analyzedArguments, isExpanded);
```

Comment:
This feels like a bunch of manual work we shouldn't be doing, rather than calling some overload of `BindInvocationExpression`.